core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java

```diff
@@ -0,0 +1,652 @@
+                }
+
+                assert presentRecords.capacity() > bucketIndex;
+                return true;
+            } else {
+                // otherwise, we need to use bitset to fill
+                if (!gapFilling) {
+                    gapFilling = true;
+
+                    // we need to get up to date, since the sampler started at unix epoch, not min timestamp
+                    if (rangeBound == RANGE_UPPER_BOUND || rangeBound == RANGE_UNBOUNDED) {
+                        bucketIndex = timestampSampler.bucketIndex(minTimestamp);
+                        nextBucketTimestamp = minTimestamp;
+                    }
+
+                    while (recordWasPresent()) {
+                        moveToNextBucket();
+                    }
+
+                    return notAtEndOfBitset();
+                }
+
+                do {
+                    moveToNextBucket();
+                } while (recordWasPresent());
+
+                return notAtEndOfBitset();
+            }
+        }
+
+        @Override
+        public long size() {
+            return -1;
+        }
+
+        @Override
+        public void toTop() {
+            baseCursor.toTop();
+            presentRecords.clear();
+            gapFilling = false;
+            baseRecord = null;
+            bucketIndex = 0;
+        }
+
+        private Function getFillFunction(int col) {
+            if (col == timestampIndex) {
+                return fillingTimestampFunc;
+            }
+            return valueFuncs.getQuick(col);
+        }
+
+        private void initBounds(Function fromFunc, Function toFunc) {
+            if (fromFunc != TimestampConstant.NULL && toFunc != TimestampConstant.NULL) {
+                rangeBound = RANGE_FULLY_BOUND;
+                return;
+            }
+
+            if (fromFunc != TimestampConstant.NULL) {
+                rangeBound = RANGE_LOWER_BOUND;
+                return;
+            }
+
+            if (toFunc != TimestampConstant.NULL) {
+                rangeBound = RANGE_UPPER_BOUND;
+                return;
+            }
+
+            rangeBound = RANGE_UNBOUNDED;
+        }
+
+        private void initTimestamps(Function fromFunc, Function toFunc, CharSequence stride) throws SqlException {
+            if (fromFunc != TimestampConstant.NULL) {
+                fromTimestamp = fromFunc.getTimestamp(null);
+            }
+
+            if (toFunc != TimestampConstant.NULL) {
+                toTimestamp = toFunc.getTimestamp(null);
+            }
+
+            timestampSampler = TimestampSamplerFactory.getInstance(stride, 0);
+            timestampSampler.setStart(fromTimestamp);
+
+            nextBucketTimestamp = fromTimestamp;
+
+            minTimestamp = fromFunc == TimestampConstant.NULL ? Long.MAX_VALUE : fromTimestamp;
+            maxTimestamp = toFunc == TimestampConstant.NULL ? Long.MIN_VALUE : toTimestamp;
+        }
+
+        private void initValueFuncs(ObjList<Function> valueFuncs) {
```

Comment:
It might be easier to insert `null` at the `timestampIndex` once, in the `FillRangeRecordCursorFactory` constructor, so that we don't need all of these idempotency checks. WDYT?

Reply:
> Moving it to the constructor breaks a dependency in the getCursor() factory, which ensures we have enough null fills.

We could fix that with a boolean field responsible for `valueFuncs` initialization and validation. To me, the `initValueFuncs` method is a bit too complicated for what it's doing, so that's why I suggested extracting the logic to the upper level.

Let's keep it as is.